metrics/chronos_httpserver.nim

```diff
@@ -55,10 +78,367 @@ when defined(metrics):
         printError(e.msg)
         sleep(1000)
 
-{.push raises: [Defect].}
+  type
+    MetricsRequest {.pure.} = enum
+      Status, Start, Stop, Close
+
+    MetricsResponse = object
+      status: byte
+      data: array[255, byte]
+
+    MetricsThreadData = object
+      reqTransp: StreamTransport
+      respTransp: StreamTransport
+      http: HttpServerRef
+
+    MetricsErrorKind {.pure.} = enum
+      Timeout, Transport, Communication
+
+  const
+    ResponseOk = 0'u8
+    ResponseError = 1'u8
+
+  proc raiseMetricsError(msg: string, exc: ref Exception) {.
+       noreturn, noinit, raises: [MetricsError].} =
+    let message = msg & ", reason: [" & $exc.name & "]: " & $exc.msg
+    raise (ref MetricsError)(msg: message, parent: exc)
+
+  proc raiseMetricsError(msg: string) {.
+       noreturn, noinit, raises: [MetricsError].} =
+    raise (ref MetricsError)(msg: msg)
+
+  proc raiseMetricsError(msg: MetricsErrorKind, exc: ref Exception) {.
+       noreturn, noinit, raises: [MetricsError].} =
+    case msg
+    of MetricsErrorKind.Timeout:
+      raiseMetricsError("Connection with metrics thread timed out", exc)
+    of MetricsErrorKind.Transport:
+      raiseMetricsError("Communication with metrics thread failed", exc)
+    of MetricsErrorKind.Communication:
+      raiseMetricsError("Communication with metrics thread failed", exc)
+
+  proc raiseMetricsError*(msg: string, err: OSErrorCode) {.
+       noreturn, noinit, raises: [MetricsError].} =
+    let message = msg & ", reason: [OSError]: (" & $int(err) & ") " &
+                  osErrorMsg(err)
+    raise (ref MetricsError)(msg: message)
+
+  proc respond(m: MetricsThreadData, mtype: byte, message: string) {.async.} =
+    var buffer: array[256, byte]
+    let length = min(len(message), len(buffer) - 1)
+    zeroMem(cast[pointer](addr buffer[0]), len(buffer))
+    buffer[0] = mtype
+    if length > 0:
+      copyMem(cast[pointer](addr buffer[1]), unsafeAddr message[0], length)
+    let res = await m.respTransp.write(cast[pointer](addr buffer[0]),
+                                       len(buffer))
+    if res != len(buffer):
+      raiseMetricsError("Incomplete response has been sent")
+
+  proc communicate(m: MetricsHttpServerRef,
+                   req: MetricsRequest): Future[MetricsResponse] {.async.} =
+    var buffer: array[256, byte]
+    buffer[0] = cast[byte](req)
+    block:
+      let res = await m.reqTransp.write(cast[pointer](addr buffer[0]), 1)
```

Comment:
Is this `cast` necessary? `ptr byte` should be convertible to `pointer` without a cast.